crawl-ref/source/timed-effects.cc

```diff
@@ -1258,3 +1259,104 @@ int speed_to_duration(int speed)
 
     return div_rand_round(100, speed);
 }
+
+static int& _zot_clock_for(branch_type br)
+{
+    static const string CLOCK_KEY_ROOT = "ZOT_CLOCK_FOR_";
+    const string branch_name = branches[br].abbrevname;
+    return you.props[CLOCK_KEY_ROOT + branch_name].get_int();
+}
+
+static int& _zot_clock() {
+    return _zot_clock_for(you.where_are_you);
+}
+
+static bool _zot_clock_active_in(branch_type br)
+{
+    return br != BRANCH_ABYSS && !player_has_orb();
+}
+
+// Is the zot clock running, or is it paused or stopped altogether?
+bool zot_clock_active() {
+    return _zot_clock_active_in(you.where_are_you);
+}
+
+static bool _over_zot_threshold(branch_type br) {
+    return _zot_clock_for(br) >= MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD;
+}
+
+// If the player was in the given branch, would they suffer penalties for
+// nearing the end of the zot clock?
+bool bezotted_in(branch_type br)
+{
+    return _zot_clock_active_in(br) && _over_zot_threshold(br);
+}
+
+// Is the player suffering penalties from nearing the end of the zot clock?
+bool bezotted() {
+    return bezotted_in(you.where_are_you);
+}
+
+// How many times should the player have been drained by Zot?
+int bezotting_level() {
+    if (!bezotted())
+        return 0;
+    const int MAX_ZOTS = 5;
+    const int TURNS_PER_ZOT = (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD) / MAX_ZOTS;
+    const int over_thresh = _zot_clock() - (MAX_ZOT_CLOCK - BEZOTTING_THRESHOLD);
+    return over_thresh / TURNS_PER_ZOT + 1;
+}
+
+// Decrease the zot clock when the player enters a new level.
+void decr_zot_clock() {
+    if (!zot_clock_active()) {
+        return;
+    }
+    if (bezotted()) {
+        mpr("As you enter the new level, Zot loses track of you.");
+    }
+    int &zot = _zot_clock();
+    zot = max(0, zot - ZOT_CLOCK_PER_FLOOR);
```

Comment:
just to double check, is it intentional that whatever remains of the delver's secret extra bonus gets wiped out when they change a level?

Reply:
It was not. I'll drop the secret extra clock for now - I think delver should be fine anyway. Fingers crossed?